src/transformers/trainer.py

```diff
@@ -1940,6 +1995,46 @@ def _inner_training_loop(
 
                 if self.control.should_epoch_stop or self.control.should_training_stop:
                     break
+
+                # Let's only materialize the weights, optimzier's moving averages for the grad, loss
+                # such that we shard both the input and output and avoid recompilation.
+                tensors = [(name, param) for name, param in model.named_parameters()]
+                optimizer_states = []
+                # Shard the optimizer, most optimizers are inited after optimizer.step()
+                for _, raw_state in self.optimizer.optimizer.state.items():
```

Comment:
Do we still need to reshard the optimizer state each step after https://github.com/pytorch/xla/pull/5508?

Reply:
I don't think so. However, I will need to run that change on larger scale.